lib/ferrum/frame/dom.rb

```diff
@@ -36,6 +36,33 @@ def body
         evaluate("document.documentElement.outerHTML")
       end
 
+      def wait_for_selector(css: nil, xpath: nil, timeout: 3000, interval: 100)
+        evaluate_func(%(
+          function(selector, isXpath, timeout, interval) {
+            var attempts = 0;
+            var max = timeout / interval;
+            function waitForSelector(resolve, reject) {
+              if (attempts > ((max < 1) ? 1 : max)) {
+                return reject(new Error("Not found element match the selector: " + selector));
+              }
+              var element = isXpath
+                ? document.evaluate(selector, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue
+                : document.querySelector(selector);
+              if (element !== null) {
+                return resolve(element);
+              }
+              setTimeout(function () {
+                waitForSelector(resolve, reject);
+              }, interval);
+              attempts++;
+            }
+            return new Promise(function (resolve, reject) {
+              waitForSelector(resolve, reject);
+            });
+          }
+        ), css || xpath, css.nil? && !xpath.nil?, timeout, interval, awaitPromise: true)
```

Comment:
I was feeling like this arguments list was a little crowded, and also noticed this function could be incorrectly called without raising an exception: one of `css` or `xpath` is required, but not handled. What do you think about something like this:

```ruby
      def wait_for_selector(css: nil, xpath: nil, timeout: 3000, interval: 100)
        raise ArgumentError.new("Provide :css or :xpath, but not both") if css && xpath
        raise ArgumentError.new("Provide :css or :xpath") unless css || xpath

        evaluate_func(<<~JS, css || xpath, !!xpath, timeout, interval, awaitPromise: true)
          function(selector, isXpath, timeout, interval) {
            var attempts = 0;
```

The guard clauses protect the arguments from being called incorrectly, and also simplify the resolution to `isXpath`, since it can now only be one or the other.

Reply:
well, it looks like too much for one method to be responsible for selector/XPath and also for validations.
However, I see the confusing point here as well. In this case, I think we should use simple methods for more clear API, like those:
```ruby
      def wait_for_selector(css, **options)
        wait_for_element(css: css, **options)
      end
      def wait_for_xpath(xpath, **options)
        wait_for_element(xpath: xpath, **options)
      end
```
The point to merge two ways of element-find logic was a follow a DRY principle for a common JS function, especially this part:
```js
              var element = isXpath
                ? document.evaluate(selector, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue
                : document.querySelector(selector);
```
therefore, I think, we can find a way to refactor it with a private function by passing a js function, gimme see closer, will do an update soon.

Reply:
Nice one, yes, I think this is a better direction.